doc/user/content/ops/deployment.md

```diff
@@ -12,61 +12,137 @@ issue](https://github.com/MaterializeInc/materialize/issues/new?labels=C-feature
+query joins multiple sources together.
+{{< /warning >}}
+
+Note that compaction is triggered in response to updates arriving. As a result,
+if updates stop arriving for a source, Materialize may never compact the source
+fully. This can result in higher-than-expected memory usage.
+
+This phenomenon is particularly evident when ingesting a source with a large
+amount of historical data (e.g, a several gigabyte Kafka topic that is no longer
+changing). With the default compaction window of 60 seconds, it is likely that
+the source will be fully ingested within the compaction window. By the time the
+data is eligible for compaction, the source is fully ingested, no new updates
+are arriving, and therefore no compaction is ever triggered.
+
+If the increased memory usage is problematic, consider one of the following
+solutions:
+
+  * Decrease the logical compaction window so that compaction is triggered while
+    the source is still ingesting.
+
+  * Compact the source upstream of Materialize.
+
+    If you are using the [upsert envelope with a Kafka
+    source](/sql/create-source/avro-kafka/#upsert-envelope-details), consider
+    setting compaction policy on the Kafka topic to have Kafka perform the
+    compaction.
+
+    If you are using a [file source](/sql/create-source/text-file), consider
+    rewriting the file to remove irrelevant historical data.
+
+  * Periodically send dummy updates to trigger compaction.
+
+### Swap
+
+To minimize the chances that Materialize runs out of memory in a production
+environment, we recommend you make additional memory available to Materialize
+via a SSD-backed swap file or swap partition.
+
+This is particularly important in Linux and in Docker, where swap may not be
+automatically set up for you.
 
 #### Docker
-[By default](https://docs.docker.com/config/containers/resource_constraints/), a
-container has no resource constraints and can use as much of a given resource as the host’s
-kernel scheduler allows, unless you have overridden this with the `--memory` or the
-`--memory-swap` flags.
+
+By default, a container has no [resource
+constraints](https://docs.docker.com/config/containers/resource_constraints/)
+and can use as much memory and swap as the host allows, unless you have
+overridden this with the `--memory` or the `--memory-swap` flags.
 
 #### Linux
-Most Linux distributions do not enable swap by default. However, you can enable it
-quite easily:
 
-1. Create a swapfile:
-   The general syntax is: `fallocate [-n] [-o offset] -l length filename`
+Most cloud Linux distributions do not enable swap by default. However, you can
+enable it quite easily with the following steps.
+
+1. Create a swapfile using the `fallocate` command.
+
+   The general syntax is: `fallocate -l <swap size> filename`. For a 1GB swap
+   file, for example:
 
-   (ie, for a 1GB swapfile):
     ```shell
-    sudo fallocate -l <swap size> /swapfile
+    $ sudo fallocate -l 1G /swapfile
```

Comment:
I have a slight preference for excluding terminal markers from code samples so they're easier to copy/pate. They're relatively easy to add via CSS and make them non-selectable, so that's something I'll gladly take on if you don't mind removing these.

Reply:
Oh sure, done!